lib/metadata/acl.js

```diff
@@ -14,10 +14,21 @@ const acl = {
 
     addObjectACL(bucket, objectKey, objectMD, addACLParams, params, log, cb) {
         log.trace('updating object acl in metadata');
-        // eslint-disable-next-line no-param-reassign
-        objectMD.acl = addACLParams;
-        metadata.putObjectMD(bucket.getName(), objectKey, objectMD, params, log,
-            cb);
+        function _didNotChange(grant) {
+            return objectMD.acl[grant].length === addACLParams[grant].length
+                && objectMD.acl[grant].every(value => addACLParams[grant].includes(value));
+        }
+        const cannedChanged = objectMD.acl.Canned !== addACLParams.Canned;
+        if (cannedChanged || !_didNotChange('FULL_CONTROL') || !_didNotChange('WRITE_ACP')
+            || !_didNotChange('READ') || !_didNotChange('READ_ACP')) {
```

Comment:
We should not hardcode a check for each grant.
If there is a list of all "allowed" grants, we iterate over it ; however, since this function does not do any such check, it would be better to just compare the keys in both "objectMD.acl" and "addACLParams" : if both don't have the same keys (whatever they may be), do the update ; else check for each key it still matches.